src/backend/distributed/planner/multi_router_planner.c

```diff
@@ -1886,11 +1889,46 @@ PlanRouterQuery(Query *originalQuery,
 
 	*placementList = NIL;
 
-	prunedRelationShardList = TargetShardIntervalsForQuery(originalQuery,
-														   plannerRestrictionContext->
-														   relationRestrictionContext,
-														   &isMultiShardQuery,
-														   partitionValueConst);
+	/*
+	 * When FastPathRouterQuery() returns true, we know that standard_planner() has
+	 * not been called. Thus, restriction information is not avaliable and we do the
+	 * shard pruning based on the distribution column in the quals of the query.
+	 */
+	if (FastPathRouterQuery(originalQuery))
+	{
+		List *shardIntervalList =
+			TargetShardIntervalForFastPathQuery(originalQuery, partitionValueConst,
+												&isMultiShardQuery);
+
+		/*
+		 * This could only happen when there is a parameter on the distribution key.
+		 * We defer error here, later the planner is forced to use a generic plan
+		 * by assigning arbitrarily high cost to the plan.
+		 */
+		if (UpdateOrDeleteQuery(originalQuery) && isMultiShardQuery)
```

Comment:
I kind of feel like we should be catching this case earlier (unresolved parameter + UPDATE/DELETE).

Reply:
What about this new commit: https://github.com/citusdata/citus/pull/2606/commits/1625c32158c9136c516707e2522058c4dde35536

The trick is that `FastPath` queries fail on `QueryPushdownSqlTaskList ()` since that function relies heavily on restrictInfos. So, I ended-up with this hack

Reply:
I like it less than the previous approach.

Reply:
So few options that I can come up:

- Previous commit so that return as early as possible 
- This commit
- Similar to this commit, but get rid of the 'else if (isMultiShardModifyQuery && FastPathRouterQuery(originalQuery))' and change 'QueryPushdownSqlTaskList' so that it doesn't rely on restrict info. Note that non-fast path router update/deletes with parameter on the dist key goes through that path on the 6th execution 

Any other option can you think of?

Reply:
I think previous commit is probably ok for now. We might want to do some refactoring in a separate PR.

Reply:
Makes sense. 

This is the first time I'm touching `dist_key = param` parts and in general I found it hard to follow. This PR is also adding one more edge to it. Already discussed on #1848 so just linking for reference.